src/town_voronoi.cpp

```diff
@@ -0,0 +1,239 @@
+/* $Id$ */
+
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file town_voronoi.cpp Handling of town Voronoi diagram, containing the TownID of the closest town for every tile. */
+
+#include "stdafx.h"
+#include "town.h"
+#include "sortlist_type.h"
+
+#include "debug.h"
+#include <chrono>
+#include <iostream>
+
+typedef GUIList<const Town*> TownList;
+bool _debug;
+bool _voronoi_initialized = false;
+
+int CDECL TownXYSorter(const Town * const *a, const Town * const *b)
+{
+	const Town *ta = *a;
+	const Town *tb = *b;
+	return (ta->xy < tb->xy) ? -1 : 1;
+}
+
+/**
+ * Calculates the maximum distance where the tiles between town A and B
+ * are closer to town A than town B. If there are tiles with equal distance
+ * from town A and B, the tile belongs to the town with the smaller TownID (index).
+ * @param ta pointer of town A
+ * @param tb pointer of town B
+ * @return the half distance between town A and B
+ */
+uint CalcHalfDistance(const Town* ta, const Town* tb)
+{
+	uint distance = DistanceManhattan(ta->xy, tb->xy);
+
+	assert(distance != 0);
+
+	if (distance % 2 == 1) return (distance -1 ) / 2;
+
+	// There are tiles that has the same distance form the two towns
+	// The dominant town (with lower index) gets the bigger part
+	return (ta->index < tb->index) ? distance / 2 : distance / 2 - 1;
+}
+
+/**
+ * From a given tile, finds the first tile (while moving in the X direction) that is 
+ * closer to town A than town B. At least one tile must be closer to town A than
+ * town B in the given 'Y map slice' for the function to work.
+ * @param ta pointer of town A
+ * @param tb pointer of town B
+ * @param limit tile index of the starting tile
+ * @return X coordinate of the found tile
```

Comment:
You should document what happens if all tiles searched are closer to B than A, and either return a sentinel value (such as 0 or UINT_MAX) or provide a method to know beforehand whether the function will be able to succeed (if that's even possible to determine).

Reply:
I made an assertion on this, so it the responsibility of the caller function to make sure there is at least one such tile.